sal/tests/server/unit/test_data.py

```diff
@@ -0,0 +1,437 @@
+
+    # Ideally we would use assert_called_with here, however as Branch and
+    # DataObject do not implement __eq__, we cannot create an object from
+    # content['object'] and test if mock_persistence_provider.put has been
+    # called with it 
+    # If Branch.__eq__ and DataObject.__eq__ are added (and they are
+    # member-wise comparisons), this should be changed
+    call_args = mock_persistence_provider.put.call_args
+    assert len(call_args) == 2
+    assert call_args[0][0] == '/{}'.format(path)
+    assert isinstance(call_args[0][1], cls)
+
+
+# Post does not currently raise an exception if a revision is supplied
+@pytest.mark.xfail(reason='Known exception omission in DataTree.post')
+@pytest.mark.parametrize('json_content, cls',
+                         [({'content':'object',
+                            'type':'branch',
+                            'object':{'description':'A branch'}},
+                            Branch),
+                          ({'content':'object',
+                            'type':'leaf',
+                            'object':{'description':{'type':'string',
+                                                     'value':'A data object'},
+                                      '_type':{'type':'string',
+                                                'value':'object'},
+                                      '_class':{'type':'string',
+                                                'value':'scalar'},
+                                      '_group':{'type':'string',
+                                                'value':'core'},
+                                      '_version':{'type':'uint64',
+                                                  'value':1},
+                                      'value':{'type':'uint64',
+                                               'value':5}}},
+                            DataObject)])
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(revision=valid_revisions)
+def test_data_tree_post_object_with_revision(revision, server,
+                                             mock_persistence_provider,
+                                             json_content, cls):
+
+    """
+    Tests that a DataTree raises an exception if a user tries to post to a path
+    with a revision number
+
+    GIVEN
+        A request with a URL path
+        AND a request with JSON content describing a branch or data object
+        AND a request with a revision=non-negative int
+    WHEN
+        The DataTree posts to the path
+    THEN
+        A BadRequest exception is raised
+    """
+
+    path = 'this/is/the/path'
+    json_content['revision'] = revision
+
+    _check_raising_request(server, 'post', path, BadRequest,
+                           json=json_content)
+
+
+# Post currently has a bug which means that the exception is returned rather
+# than raised, so this is currently an expected fail
+@pytest.mark.xfail(reason='Known bug in DataTree.post')
```

Comment:
**praise:** Documenting bugs by having xfailing tests is a great idea, as it also reduces the preliminary work required when fixing them.

Reply:
Thanks - yeah if you look at the number of issues created in the last few weeks you'll see there are quite a few xfails around!